text/0000-target-feature-trait-methods.md

```diff
@@ -0,0 +1,251 @@
+
+# Motivation
+[motivation]: #motivation
+
+Imagine we're writing an AVX library with the goal of being generic over f32 and f64.  We can start by creating a trait
+called `AvxVector` with some methods common to both f32 and f64 vectors:
+
+```rust
+pub trait AvxVector {
+    unsafe fn add(left: Self, right: Self) -> Self;
+}
+impl AvxVector for __m256 {
+    unsafe fn add(left: Self, right: Self) -> Self {
+        _mm256_add_ps(left, right)
+    }
+}
+impl AvxVector for __m256d {
+    unsafe fn add(left: Self, right: Self) -> Self {
+        _mm256_add_pd(left, right)
+    }
+}
+```
+The `RustFFT` crate implements a [real-world example of this](https://github.com/ejmahler/RustFFT/blob/simd/src/avx/avx_vector.rs#L16)
+
+Now, users of our library can compute a sum of f32 and f64 vectors with the same code:
+```rust
+#[target_feature(enable = "avx")]
+unsafe fn do_work() {
+    let left32 : __m256 = ...;
+    let right32 : __m256 = ...;
+    let sum32 = unsafe { AvxVector::add(left32, right32) }
+
+    let left64 : __m256d = ...;
+    let right64 : __m256d = ...;
+    let sum64 = unsafe { AvxVector::add(left64, right64) }
+}
+```
+
+Everything so far compiles on stable Rust, but there's a major flaw: Despite every single function being marked unsafe, absolutely nothing unsafe is happening here! [RFC 2396 (`target feature 1.1`)] improves the situation for standalone functions by allowing calls to a `#[target_feature(..)]` function to be safe as long as the caller also has `#[target_feature(..)]` with the same features.
+
+However, that change doesn't apply here, because `#[target_feature(..)]` isn't allowed on trait methods. If it was allowed, this example could be written with no unsafe whatsoever.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Currently, we don't allow `#[target_feature]` to appear on trait methods:
+
+```rust
+pub trait AvxVector {
+    #[target_feature(enable = "avx")]
+    fn add(left: Self, right: Self) -> Self;
+}
+```
+
+```rust
+error: attribute should be applied to a function
+ --> src\lib.rs:6:5
+  |
+6 |     #[target_feature(enable = "avx")]
+  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+7 |     fn add(left: Self, right: Self) -> Self;
+  |     ---------------------------------------- not a function
+```
+
+With this RFC, the error above is no longer an error, and the trait definition comples successfully. To ensure consistency between trait definitions and trait impls, we require that the `#[target_feature]` declaration is identical between the trait definition and the trait impl. Some examples:
+
+```rust
+pub trait AvxVector {
+    #[target_feature(enable = "avx")]
+    fn add(left: Self, right: Self) -> Self;
+
+    fn sub(left: Self, right: Self) -> Self;
+
+    #[target_feature(enable = "avx")]
+    fn mul(left: Self, right: Self) -> Self;
+
+    #[target_feature(enable = "avx")]
+    fn div(left: Self, right: Self) -> Self;
+}
```

Comment:
Lets say that you wanted all these methods to be `#[target_feature(enable = "avx")]`. Should it be possible to mark the `trait` itself as such? Perhaps this could be an "alternative" or "future work".

```rust
#[target_feature(enable = "avx")]
pub trait AvxVector {
    fn add(left: Self, right: Self) -> Self;

    fn sub(left: Self, right: Self) -> Self;

    fn mul(left: Self, right: Self) -> Self;
    
    fn div(left: Self, right: Self) -> Self;
}
```